rtos/TARGET_CORTEX/mbed_rtx_idle.cpp

```diff
@@ -95,21 +95,22 @@ uint32_t OS_Tick_GetInterval (void) {
 
 static void default_idle_hook(void)
 {
-    uint32_t elapsed_ticks = 0;
+    uint32_t ticks_to_sleep = osKernelSuspend();
+    os_timer->suspend(ticks_to_sleep);
 
-    core_util_critical_section_enter();
-    uint32_t ticks_to_sleep = svcRtxKernelSuspend();
-    if (ticks_to_sleep) {
-        os_timer->schedule_tick(ticks_to_sleep);
+    while (!os_timer->suspend_time_passed()) {
 
-        sleep();
+        core_util_critical_section_enter();
+        if (!osRtxInfo.kernel.pendSV) { // check devices didn't run any interrupt handlers kicking the OS between suspend and critical section
+            sleep();                    // uses WFI - will wake as soon as any pending interrupts
+        }
+        core_util_critical_section_exit();
 
-        os_timer->cancel_tick();
-        // calculate how long we slept
-        elapsed_ticks = os_timer->update_tick();
+        if (osRtxInfo.kernel.pendSV) {
```

Comment:
This could just be part of the while condition? Could be true on first entry to the while